source/extensions/transport_sockets/alts/tsi_handshaker.h

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/common/c_smart_ptr.h"
+
+#include "extensions/transport_sockets/alts/grpc_tsi.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+/**
+ * An interface to get callback from TsiHandshaker. TsiHandshaker will call this callback in the
+ * thread which its dispatcher posts to.
+ */
+class TsiHandshakerCallbacks {
+public:
+  virtual ~TsiHandshakerCallbacks() {}
+
+  struct NextResult {
+    // A enum of the result
+    tsi_result status_;
+
+    // The buffer to be sent to the peer
+    Buffer::InstancePtr to_send_;
+
+    // A pointer to tsi_handshaker_result struct. Owned by instance.
+    CHandshakerResultPtr result_;
+  };
+
+  typedef std::unique_ptr<NextResult> NextResultPtr;
+
+  /**
+   * Called when `next` is done, this may be called inline in `next` if the handshaker is not
+   * asynchronous.
+   * @param result
+   */
+  virtual void onNextDone(NextResultPtr&& result) PURE;
+};
+
+/**
+ * A C++ wrapper for tsi_handshaker interface.
+ * For detail of tsi_handshaker, see
+ * https://github.com/grpc/grpc/blob/v1.12.0/src/core/tsi/transport_security_interface.h#L236
+ */
+class TsiHandshaker final : public Event::DeferredDeletable {
+public:
+  explicit TsiHandshaker(CHandshakerPtr&& handshaker, Event::Dispatcher& dispatcher);
+  ~TsiHandshaker();
+
+  /**
+   * Conduct next step of handshake, see
+   * https://github.com/grpc/grpc/blob/v1.10.0/src/core/tsi/transport_security_interface.h#L416
+   * It is callers responsibility to not call this method again until the onNextDone is called.
+   * @param received the buffer received from peer.
+   */
+  tsi_result next(Buffer::Instance& received);
+
+  /**
+   * Set handshaker callbacks, this must be called before calling next.
```

Comment:
callbacks.  This must. 

Alternately would it make sense to create the handshaker with the callbacks?